### ISSUE TITLE
Link two flycameras

### ISSUE THREAD

**Issue by SuyiWang** [vispy/vispy#1554]
The mouse rotation for two linked FlyCameras like the following is not working properly.
```
    vb1 = scene.widgets.ViewBox(border_color='yellow', parent=canvas.scene)
    vb2 = scene.widgets.ViewBox(border_color='blue', parent=canvas.scene)
    # Creates flycamera for two side-by-side scene. Setup steps omitted.
    camfly1, camfly2 = FlyCamera(), FlyCamera()
    vb1.camera = camfly1
    vb2.camera = camfly2
    vb1.camera.link(vb2.camera)
```
Namely, the current mouse rotation for linked FlyCameras has an inertia effect. For example, let's say we have two movements with left button down. Move 1: Dragging the screen from `mouse_down_pos` to `mouse_pos_A`. Move 2: Without releasing left mouse button, dragging the screen from `mouse_pos_A` back to `mouse_down_pos`. Without linked camera, the FlyCamera always looks at where mouse points at, which means it starts to look back to `mouse_down_pos` in Move 2. However, with FlyCameras linked, those cameras never look back, unless one keeps Move 2 and moves mouse pass `mouse_down_pos`.

I also noticed that at Line 55-56
```
    # Linking this camera likely not to work very well
    _state_props = PerspectiveCamera._state_props + ('rotation', )
```

I have found that this issue is caused by how set_state is implemented (BaseCamera.py, Line 322 ). It copies all properties from one camera to another. 
```
        for key, val in D.items():
            if key not in self._state_props:
                raise KeyError('Not a valid camera state property %r' % key)
            setattr(self, key, val)
```
However in FlyCamera.py the `rotation` property is: 
```
    @property
    def rotation(self):
        """ Get the full rotation. This rotation is composed of the
        normal rotation plus the extra rotation due to the current
        interaction of the user.
        """
        rotation = self._rotation2 * self._rotation1
        return rotation.normalize()

    @rotation.setter
    def rotation(self, value):
        assert isinstance(value, Quaternion)
        self._rotation1 = value
```
So what happened for linked camera is that it reads rotation2*rotation1 as rotation from one camera and writes to rotation1 in another. This rotation1 from the other camera will later affect rotation1 in current camera.
I have a quick fix in FlyCamera class, which asks set_state to copy _rotation1 and _rotation2 instead of rotation property.
```
    # In Fly.py, class FlyCamera()
    _state_props = PerspectiveCamera._state_props + ('rotation_self', 'rotation_user')
    @property
    def rotation(self):
        rotation = self._rotation2 * self._rotation1
        return rotation.normalize()
    
    @rotation.setter
    def rotation(self, value):
        print("called rotation.setter, nothing done")

    @property
    def rotation_self(self):
        return self._rotation1.normalize()
    
    @rotation_self.setter
    def rotation_self(self, value):
        assert isinstance(value, Quaternion)
        self._rotation1 = value

    @property
    def rotation_user(self):
        return self._rotation2.normalize()
    
    @rotation_user.setter
    def rotation_user(self, value):
        assert isinstance(value, Quaternion)
        self._rotation2 = value
```
This works on my application, but I would like to know if this might cause any problem, since the rotation property might be used somewhere else in vispy. In my fix, one should use `flycamera.rotation_self` setter whenever one wants to use `flycamera.rotation` setter in before.
At the least, I hope the cause of the issue is made clear and we can come up with a better fix.

**Comment by djhoese** [comment]
Are you following an existing example that shows this issue? Otherwise do you have a self-contained example that I could test out? What backend are you using?

Thanks for the thorough bug report too.

**Comment by SuyiWang** [comment]
> Are you following an existing example that shows this issue? Otherwise do you have a self-contained example that I could test out? What backend are you using?
> 
> Thanks for the thorough bug report too.

Here is an example script to reproduce the bug. If you unquote the block in `class fly`, the issue should be gone.
```
import numpy as np
from vispy import app, scene, io
from vispy.scene.cameras.perspective import PerspectiveCamera
from vispy.util.quaternion import Quaternion

class fly(scene.FlyCamera):
    pass
    '''
    _state_props = PerspectiveCamera._state_props + ('rotation_self', 'rotation_user')
    @property
    def rotation(self):
        rotation = self._rotation2 * self._rotation1
        return rotation.normalize()
    
    @rotation.setter
    def rotation(self, value):
        print("called rotation.setter, nothing done")

    @property
    def rotation_self(self):
        return self._rotation1.normalize()
    
    @rotation_self.setter
    def rotation_self(self, value):
        assert isinstance(value, Quaternion)
        self._rotation1 = value

    @property
    def rotation_user(self):
        return self._rotation2.normalize()
    
    @rotation_user.setter
    def rotation_user(self, value):
        assert isinstance(value, Quaternion)
        self._rotation2 = value
    '''

canvas = scene.SceneCanvas(keys='interactive')
canvas.size = 900, 400
canvas.show()

# Create two ViewBoxes, place side-by-side
vb1 = scene.widgets.ViewBox(border_color='yellow', parent=canvas.scene)
vb2 = scene.widgets.ViewBox(border_color='blue', parent=canvas.scene)
#
grid = canvas.central_widget.add_grid()
grid.padding = 6
grid.add_widget(vb1, 0, 0)
grid.add_widget(vb2, 0, 1)

# Create the image
cube1 = scene.visuals.Cube(size=1, color='red', edge_color="black",
                          parent=vb1.scene)
cube2 = scene.visuals.Cube(size=1, color='yellow', edge_color="black",
                          parent=vb2.scene)
# setup camera
vb1.camera, vb2.camera = fly(), fly()
vb1.camera.link(vb2.camera)

if __name__ == '__main__':
    app.run()
```
I don't quite get what "backend" refers to in this context, but my python version is 3.5.2 and the OS is Ubuntu 16.04.4 LTS. Would you elaborate if these two guesses are not helping?

**Comment by djhoese** [comment]
The backend refers to the GUI framework that vispy is using to draw things on the screen, like PyQt5, PyQt4, Wx, etc. You can find this type of information out by doing:

```python
import vispy
print(vispy.sys_info())
```

If you paste the output of that here that may give me more information, but I'm guessing you are correct about limitations of the current implementation of the fly cameras.

**Comment by djhoese** [comment]
I've looked at the fly camera and your original post and I think I understand what you're saying now. As for your concerns about changing rotation, I don't think it is used anywhere else in vispy. Is there a reason you called them `_self` and `_user`? And what is the reason for the `normalize` call inside the setter methods?

Regardless, I think it would be great if you could make a pull request with these changes so the other vispy devs and I can properly review it and get it in to vispy.

**Comment by SuyiWang** [comment]
It seems that my reply was not posted. Just post it again. [The backend is PyQt5.]

> Is there a reason you called them `_self` and `_user`? 
The original idea is to let people understand what the variables mean by their names. I am also considering names such as `_fixed`,`_varying`. It could also be good to just put comments below `_rotation1` and `_rotation2`.

> And what is the reason for the `normalize` call inside the setter methods?
My version has no `normalize` for setter, but for properties. In original version, the `rotation` property returns a normalized quaterion so I did it for both `rotation1` and `rotation2`. This looks reasonable to me because that rotation1 and rotation2 may not be normalized in setter.

Sure, I am making the pull request now.

**Comment by SuyiWang** [comment]
Now there is a small issue left. Let's still use the following example. The snippet is using FlyCamera after the commit above.
```
import numpy as np
from vispy import app, scene, io

canvas = scene.SceneCanvas(keys='interactive')
canvas.size = 900, 400
canvas.show()

# Create two ViewBoxes, place side-by-side
vb1 = scene.widgets.ViewBox(border_color='yellow', parent=canvas.scene)
vb2 = scene.widgets.ViewBox(border_color='blue', parent=canvas.scene)
#
grid = canvas.central_widget.add_grid()
grid.padding = 6
grid.add_widget(vb1, 0, 0)
grid.add_widget(vb2, 0, 1)

# Create the image
cube1 = scene.visuals.Cube(size=1, color='red', edge_color="black",
                          parent=vb1.scene)
cube2 = scene.visuals.Cube(size=1, color='yellow', edge_color="black",
                          parent=vb2.scene)
# setup camera
vb1.camera, vb2.camera = scene.FlyCamera(name='fly1'), scene.FlyCamera(name='fly2')
vb2.camera.link(vb1.camera)
# vb1.interactive = False

if __name__ == '__main__':
    app.run()
```
When started new, there is no problem in control both cameras in either side of the viewbox. But after resizing the window / moving cameras using keyboard / rotating cameras using mouse in viewbox1 (left), the mouse rotation for camera in viewbox2 (right) becomes very ir-responsive (aka. frame drop). And camera control for viewbox1 never becomes ir-responsive.

**Comment by SuyiWang** [comment]
The small issue mentioned above could be resolved by adding `event.handled = True` in several locations. The change is reflected in the second commit. Afaik, there is no issue left.

**Comment by djhoese** [comment]
Closed by #1557
